book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,220 @@
+
+```
+pub struct SpendV4 {
+    pub cv: commitment::ValueCommitment,
+    pub anchor: tree::Root,
+    pub nullifier: note::Nullifier,
+    pub rk: redjubjub::VerificationKeyBytes<SpendAuth>,
+    pub zkproof: Groth16Proof,
+    pub spend_auth_sig: redjubjub::Signature<SpendAuth>,
+}
+```
+
+The output code is located at `zebra-chain/src/sapling/output.rs`:
+
+```
+pub struct OutputV4 {
+    pub cv: commitment::ValueCommitment,
+    pub cm_u: jubjub::Fq,
+    pub ephemeral_key: keys::EphemeralPublicKey,
+    pub enc_ciphertext: note::EncryptedNote,
+    pub out_ciphertext: note::WrappedNoteKey,
+    pub zkproof: Groth16Proof,
+}
+```
+
+Now lets see how the v5 transaction is specified in the protocol, this is the second table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus).
+
+We propose the following representation for transaction V5 in Zebra:
+
+```
+V5 {
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    sapling_value_balance: Amount,
+    sapling_anchor: tree::Root,
+    sapling_shielded_data: Option<SaplingShieldedDataV5>,
+    orchard_flags: OrchardFlags,
+    orchard_value_balance: Amount,
+    orchard_anchor: tree::Root,
+    orchard_shielded_data: Option<OrchardShieldedData>,
+}
+```
+
+`SaplingShieldedDataV5` is different from `SaplingShieldedDataV4` so a new type will be defined and implemented in `zebra-chain/src/transaction/shielded_data.rs`.
+Definition will look as follows:
+
+```
+pub struct SaplingShieldedDataV5 {
+    pub first: Either<SpendV5, OutputV5>,
+    pub rest_spends: Vec<SpendV5>,
+    pub rest_outputs: Vec<OutputV5>,
+    pub binding_sig: Signature<Binding>,
+}
+```
+
+This leads to new sapling output and spend types.
+
+V5 sapling spend will be defined and implemented in `zebra-chain/src/sapling/spend.rs` (this is in the same file as `SpendV4`):
+```
+pub struct SpendV5 {
+    pub cv: commitment::ValueCommitment,
+    pub nullifier: note::Nullifier,
+    pub rk: redjubjub::VerificationKeyBytes<SpendAuth>,
+    pub zkproof: Groth16Proof,
+    pub spend_auth_sig: redjubjub::Signature<SpendAuth>,
+}
+```
+
+V5 sapling output will live in `zebra-chain/src/sapling/output.rs` (this is in the same file as `OutputV4`):
+
+```
+pub struct OutputV5 {
+    pub cv: commitment::ValueCommitment,
+    pub cm_u: jubjub::Fq,
+    pub ephemeral_key: keys::EphemeralPublicKey,
+    pub enc_ciphertext: note::EncryptedNote,
+    pub out_ciphertext: note::WrappedNoteKey,
+    pub zkproof: Groth16Proof,
+}
+```
+
+The new V5 structure will create a new `OrchardShieldedData` type. This new type will be defined in a separated file: `zebra-chain/src/transaction/orchard_shielded_data.rs` and it will look as follows:
+```
+pub struct OrchardShieldedData {
+    pub first: Action,
+    pub rest_spends: Vec<Action>,
+    pub binding_sig: Signature<Binding>,
```

Comment:
@dconnolly is the Orchard binding signature type the same as Sapling?

Reply:
Orchard binding sigs will be of a different type or at least have different Rust constraints as they use different curves. This will probably use `redpallas::Signature<redpallas::Binding>` or similar depending where we pull it from.

Reply:
@oxarbitrage can you make these changes?